### PR TITLE
Exception when uploading instances

### LINKED ISSUE #975: Exception when uploading instances
#### Software and hardware versions 
Collect latest from master on Moto X Pure

#### Problem description
When uploading instance, there's exception when closing the response stream. The upload is successful though.

#### Steps to reproduce the problem
Upload a single instance to the nafundi test server (or any aggregate server).

#### Expected behavior
No stacktrace in the log.

#### Other information

    04-25 16:11:42.955 19488-26186/org.odk.collect.android I/InstanceUploaderTask: Using Uri remap for submission 104. Now: https://nafundi-test.appspot.com/submission
    04-25 16:11:42.956 19488-26186/org.odk.collect.android I/InstanceUploaderTask: added xml_submission_file: basic_2017-03-06_10-39-48_1090.xml
    04-25 16:11:42.957 19488-26186/org.odk.collect.android I/InstanceUploaderTask: Issuing POST request for 104 to: https://nafundi-test.appspot.com/submission
    04-25 16:11:43.548 19488-26186/org.odk.collect.android E/WebUtils: Unable read the stream
    java.io.IOException: Attempted read from closed stream.
        at org.opendatakit.httpclientandroidlib.impl.io.ContentLengthInputStream.read(ContentLengthInputStream.java:167)
        at org.opendatakit.httpclientandroidlib.conn.EofSensorInputStream.read(EofSensorInputStream.java:137)
        at libcore.io.Streams.skipByReading(Streams.java:159)
        at java.io.InputStream.skip(InputStream.java:229)
        at org.odk.collect.android.utilities.WebUtils.discardEntityBytes(WebUtils.java:290)
        at org.odk.collect.android.tasks.InstanceUploaderTask.uploadOneSubmission(InstanceUploaderTask.java:463)
        at org.odk.collect.android.tasks.InstanceUploaderTask.processChunk(InstanceUploaderTask.java:575)
        at org.odk.collect.android.tasks.InstanceUploaderTask.doInBackground(InstanceUploaderTask.java:599)
        at org.odk.collect.android.tasks.InstanceUploaderTask.doInBackground(InstanceUploaderTask.java:74)
        at android.os.AsyncTask$2.call(AsyncTask.java:295)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
        at java.lang.Thread.run(Thread.java:818)

### PR DESCRIPTION
**Pull request by grzesiek2010**
This PR is in reference to resolve #975

### CODE REVIEW

**Comment by lognaturel** [comment]
See #975 for a discussion on the approach. If you know these APIs, please chime in!

**Comment by alxndrsn** [comment]
At a glance, this code doesn't look any safer than the original:

1. checking `available()` is not a reliable way to know if there are bytes available in the stream - as linked by @grzesiek2010 in the related issue:

https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#available():
> It is never correct to use the return value of this method to allocate a buffer intended to hold all data in this stream. 


2. the code still makes the assumption that `skip()` will always skip as many bytes as possible:

https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#skip(long):
> The skip method may, for a variety of reasons, end up skipping over some smaller number of bytes, possibly 0.

**Comment by alxndrsn** [comment]
And:

3. `is` is still not being closed reliably

Which I understood to be the underlying problem.

**Comment by lognaturel** [comment]
Thanks so much for checking on this. The reason we were looking at it is because someone noticed that an exception was being logged after every single successful send! Doing a debug run looking at the code that skipped bytes let me see that the concrete type of the `InputStream` is `EofSensorInputStream` and that it was being closed and then `skip` was being called. So that was the source of the crash.

I still don't really understand `discardEntityBytes` or why it was there in the first place but the comments suggest that it's supposed to handle a bad server case (I don't know what kind of bad server, though). It does seem like in the common case, `is` is empty. So as I understand it, because the call on `skip` is in the while loop condition, it would crash. Doing the check on byte availability makes sure that crash doesn't happen. The way I understood the description of `available` is that 0 means there's nothing in the stream but a non-zero value is only an approximation.

What is a case in which `is` isn't closed reliably? As I understand it, if the stream is empty, it goes right down to the close. If it has something in, we're back to where we were before where it will close because it's an `EofSensorInputStream` and will hit the exception.

Basically, this leaves the code mostly intact but prevents the exception throwing in the common case. If the exception is thrown in some other context, it will be logged to Firebase Crash and might give us a hint as to what all this is for. Unless you have a quick fix you're confident in or I've gotten something totally wrong and we've introduced new bad behavior, I'm inclined not to spend too much more time on this for now!!

**Comment by alxndrsn** [comment]
New PR at https://github.com/opendatakit/collect/pull/1103

**Comment by alxndrsn** [comment]
@lognaturel I hope this PR helps.

Some notes:

> it was being closed and then `skip` was being called. So that was the source of the crash.

Checking `.available()` on a closed `InputStream` could itself be throwing an exception, so I don't think the accepted change would necessarily have fixed the bug.


> The way I understood the description of available is that 0 means there's nothing in the stream but a non-zero value is only an approximation.

My reading is that `0` implies (but does not dictate) that there is _currently_ no data available in the stream.  However, this may just be because the stream source (e.g. network request) is being read more slowly than the stream is being processed.


> What is a case in which is isn't closed reliably?

If either `available()` or `skip()` throws an exception, `is.close()` is never called.


> As I understand it, if the stream is empty, it goes right down to the close.

Yes, this looks to be true.

Also worth nothing that the `skip()` code as implemented is not guaranteed to read the entire stream.
